src/electron/https.ts

```diff
@@ -0,0 +1,139 @@
+// Copyright 2021 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as http from 'http';
+import * as https from 'https';
+import * as tls from 'tls';
+
+import {HttpsRequest, HttpsResponse} from '../www/app/net';
+import * as errors from '../www/model/errors';
+
+const HTTPS_TIMEOUT_MS = 30 * 1000;
+
+/**
+ * Retrieves data over HTTPs. When `request.certificateFingerprint` is set, validates the server TLS
+ * certificate by comparing its fingerprint to the trusted certificate fingerprint.
+ *
+ * @param {HttpsRequest} request - HTTP request to send.
+ * @returns {Promise<HtttpResponse>} - server HTTP response.
+ * @throws {InvalidHttpsUrlError} - if `req.url` is not a valid HTTPs URL.
+ * @throws {DomainResolutionError} - if the request host cannot be resolved.
+ * @throws {CertificateValidationError} - if server certificate cannot be validated.
+ * @throws {ConnectionTimeout} - if the connection times out.
+ * @throws {ConnectionError} - if there are errors during the connection.
+ * @throws {UnexpectedPluginError} - for other error condidtions.
+ */
+export async function fetchHttps(request: HttpsRequest): Promise<HttpsResponse> {
+  let url: URL;
+  try {
+    url = new URL(request.url);
+  } catch (e) {
+    throw new errors.InvalidHttpsUrlError('failed to parse request URL');
+  }
+  if (url.protocol !== 'https:') {
+    throw new errors.InvalidHttpsUrlError('protocol must be https');
+  }
+  const validateServerCertificate = !!request.hexSha256CertFingerprint;
```

Comment:
Nit: we only ever use `!validateServerCertificate`, so maybe this could be `invalidServerCertificate` instead to slightly simplify